Adds the ability to white/blacklist jobs, branches and ranks per species and map

PsiOmegaDelta opened the pull request:
By default Vox may only join as AI or cyborg. For Torch the merchant and stowaway roles are also added.
Makes job restriction checks more re-used, rather than copy-pasted all over the place.

Also adds the missing boldannounce span class.

Also added branch and rank restrictions. For the current setup see https://github.com/Baystation12/Baystation12/pull/17896#issuecomment-316495502

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

PsiOmegaDelta committed:
Adds the ability to white/blacklist jobs per species and map

By default Vox may only join as AI or cyborg. For Torch the merchant and stowaway roles are also added.
Makes job restriction checks more re-used, rather than copy-pasted all over the place.

Also adds the missing boldannounce span class.

PsiOmegaDelta commented:
I'm willing to take suggestions for other species to job restrictions.

Coureere commented:
>I'm willing to take suggestions for other species to job restrictions.

Restrict all other xenos to civilian jobs and EC low ranks when necessary.

PsiOmegaDelta commented:
What does "when necessary" mean?

mustafakalash commented:
I assume he means restrict them according to [this list](https://baystation12.net/forums/threads/change-in-xeno-policy.4868/). Additionally, I believe GAS are to be initially restricted to passenger? @FTangSteve I also believe Diona may only play as RD with special permission, but that should probably not include a mechanical enforcement for simplicity.

PsiOmegaDelta commented:
I suppose I should put in work to restrict branches and ranks per species as well then

PurpleMartinJCK commented:
Not passenger, but giant armoured serpentids should be restricted to janitor and whatever research assistant is to start. I can expand it as needed when I release more jobs for them.

Coureere commented:
>What does "when necessary" mean?

Sorry for the late responce, i meant exactly what mustafakalash stated.

PsiOmegaDelta committed:
Adds branch and rank restrictions per species and map.

Made it a best effort implementation of https://baystation12.net/forums/threads/change-in-xeno-policy.4868/.

PsiOmegaDelta commented:
Could use some assistance with the setup, specifically allowed/restricted roles, so here is the current one for Torch specifically:
Branches, from a coding perspective:
* Expeditionary Corps (EC)
* Fleet
* Marine Corps (MC)
* Civilian

Species restrictions:
* Humans:
  * None
* Tajara, Skrell:
  * May only select the EC and Civilian branches.
  * May only hold rank E1 to E4 as well as 01 in the EC branch.
  * May hold all Civilian 'ranks'.
  * May not select SCG security or command roles.
* IPCs
  * May select any branch except for the MC.
  * May only hold rank E1 to E5 as well as 01 in the EC branch.
  * May only hold rank E1 to E5 as well as 01 in the Fleet branch.
  * May hold all Civilian 'ranks'.
  * May not select SCG security or command roles.
* Diona, Unathi
  * May only select the Civilian branch.
  * May hold all Civilian 'ranks'.
  * May not select SCG security or command roles.
* Nabber/Giant Armoured Serpentids
  * May only select the Civilian branch.
  * May hold all Civilian 'ranks'.
  * May only select the AI, cyborg, janitor, and research assistant roles.
* Vox
  * May not select any branch.
  * May not hold any rank.
  * May only select the AI, cyborg, merchant, and stowaway roles.

PurpleMartinJCK commented:
Is it possible to remove merchant and stowaway from nabbers?  Or is that still admin needed and so doesn't matter too much?

PsiOmegaDelta commented:
Adjusted.

PurpleMartinJCK commented:
Thanks Psi!

PsiOmegaDelta committed:
Random job assignment now respects job restrictions.

This includes species, branch, and rank checks.

PsiOmegaDelta commented:
I ruined the review by adding a missing check for random job assignment, which for the longest time never checked branch and rank restrictions.
It now uses the handy `job/is_restricted()` proc.

mustafakalash commented:
Can we get a final call from @ThatOneGuy42 or @Spookerton that the configuration is kosher?

Techhead0 reviewed (dismissed):
Rereviewed.

PurpleMartinJCK commented:
Oh, dionaea shouldn't be allowed to be security guard. 

And on the forum it says that Taj can't ever be research director. 
Unathi can't be corporate liason.

PsiOmegaDelta commented:
I suppose I expected too much when I hoped the main post would actually be updated.

PsiOmegaDelta committed:
Adds more species role restrictions.

IPC, Tajara, Skrell, and Unahti are now restricted from SCG security and command roles, even if they are allowed the appropriate branch are rank.

PsiOmegaDelta commented:
Now programmatically blacklists IPCs, tajara, skrell and unathi from roles according to the following:
* It is a security and/or command role.
* It belongs to any branch.
  * Unless any of the those branches is the civilian branch, in which case the role is allowed anyway.
  * I.e. it is a pure SCG role (until someone adds more branches, in which case an update will be needed).

With this the restrictions should fully match the original post at https://baystation12.net/forums/threads/change-in-xeno-policy.4868/
If you desire further restrictions, please ask for the post to the updated accordingly.
